Kernel/Devices/PIT.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Devices/PIT.h>
+#include <Kernel/Interrupts/PIC.h>
+#include <Kernel/Scheduler.h>
+#include <Kernel/Thread.h>
+#include <LibBareMetal/IO.h>
+
+#define IRQ_TIMER 0
+namespace Kernel {
+
+//#define PIT_DEBUG
+
+static PIT* s_the;
+
+void PIT::initialize()
+{
+    if (s_the == nullptr) {
+        s_the = new PIT();
+    }
+}
+
+PIT& PIT::the()
+{
+    if (s_the == nullptr) {
```

Comment:
Might be missing something here, but isn't s_the always set in initialize?

Reply:
Well, I can replace this with assertion. The idea is that if someone call the PIT singleton class before initializing it, it won't fail.

Reply:
Yeah fair enough. I just saw similar code between the() and initialize() -- and the pattern looked slightly different to me from what I was used to for whatever reason :)